src/iso.c

```diff
@@ -0,0 +1,494 @@
+            return -1;
+        } else {
+            //Do not start at the start of the buffer, because that is the start of
+            //the current extent, not where the cur_extent + cur_offset is
+            memcpy(dest, buffer + stream->cur_offset, elem_size * num_elem);
+
+            //copy the last block read in into the global "cache"
+            memcpy(global_atapi_block, buffer + (ATAPI_BLOCKSIZE * (atapi_blocks_to_read - 1)), ATAPI_BLOCKSIZE);
+
+            //update stream, as iso_seek is a mock call to keep track of
+            //"where we are" on the ISO image and does not actually move
+            //the the CD-reading head to read the ISO image, atapi_read does that
+            iso_media_seek(stream, bytes_needed, SEEK_CUR);
+            global_atapi_unit = stream->ata_unit;
+            global_atapi_extent = stream->cur_extent;
+            return num_elem;
+        }
+    } else {
+        //Do not start at the start of the buffer, because that is the start of
+        //the current extent, not where the cur_extent + cur_offset is
+        memcpy(dest, global_atapi_block + stream->cur_offset, elem_size * num_elem);
+
+        //update stream, as iso_seek is a mock call to keep track of
+        //"where we are" on the ISO image and does not actually move
+        //the the CD-reading head to read the ISO image, atapi_read does that
+        iso_media_seek(stream, bytes_needed, SEEK_CUR);
+        return num_elem;
+    }
+}
+
+/**
+ * @brief Moves the iso_point
+ * @details Changes the offset and extent values of the iso_point as instructed
+ *
+ * @param iso_p Pointer to the iso_point to adjust
+ * @param offset How far to move the current pointer
+ * @param whence The relative place to move the pointer from, SEEK_CUR is moving forward from current position, SEEK_SET is from the zeroth byte of the iso file
+ */
+static void iso_media_seek(struct iso_point *iso_p, long offset, int whence) {
+    switch(whence) {
+        case SEEK_SET:
+            //from the start of the file
+            iso_p->cur_extent = offset / ATAPI_BLOCKSIZE;
+            iso_p->cur_offset = offset % ATAPI_BLOCKSIZE;
+            break;
+        case SEEK_CUR:
+            //from the current position
+            iso_p->cur_extent += offset / ATAPI_BLOCKSIZE;
+            iso_p->cur_offset += offset % ATAPI_BLOCKSIZE;
+
+            //Check if this caused a wrap around, and if so deal with it
+            if (iso_p->cur_offset > ATAPI_BLOCKSIZE) {
+                iso_p->cur_offset -= ATAPI_BLOCKSIZE;
+                iso_p->cur_extent++;
+            }
+            break;
+        default:
+            console_printf("Illegal iso_seek mode\n");
+            break;
+    }
+}
+
+/**
+ * @brief Lookup offset the dir/file requested
+ * @details Finds the offset of the file or directory named by
+ * the full pathname starting at the root of the iso filesystem
+ * (no relative paths allowed).
```

Comment:
Does this restriction then also apply to iso_dir, which uses this function? If so, we should make sure to document it there as well.